opacus/layers/dp_lstm.py

```diff
@@ -1,11 +1,55 @@
 import math
-from typing import List, Optional, Tuple
+from typing import List, Optional, Tuple, Union
 
 import torch
 import torch.nn as nn
+from torch.nn.utils.rnn import PackedSequence
 
 from .param_rename import ParamRenamedModule
 
+from torch.nn.utils.rnn import pad_sequence, pack_padded_sequence
+
+
+def compute_seq_lengths(batch_sizes: torch.Tensor):
+    max_batch_size = batch_sizes[0]
+    if len(batch_sizes) == 1:
+        return [1] * max_batch_size
+
+    running_seq = 0
+    running_seq_lengths = []
+    for i in range(1, len(batch_sizes)):
+        delta = batch_sizes[i - 1].item() - batch_sizes[i].item()
+        running_seq += 1
+        running_seq_lengths += delta * [running_seq]
+
+    running_seq += 1
+    running_seq_lengths += batch_sizes[-1].item() * [running_seq]
+    running_seq_lengths.reverse()
+    return running_seq_lengths
+
+
+def compute_last_states(h_n: torch.Tensor, c_n: torch.Tensor, seq_lengths: List):
+    max_batch_size = len(seq_lengths)
+    hidden_size = h_n[0].shape[-1]
+    h_last = torch.zeros(max_batch_size, hidden_size)
+    c_last = torch.zeros(max_batch_size, hidden_size)
+
+    for i, seq_len in enumerate(seq_lengths):
+        h_last[i, :] = h_n[seq_len - 1][i, :]
+        c_last[i, :] = c_n[seq_len - 1][i, :]
+
+    return h_last, c_last
+
+
+def concat_tensor_sequence(a: Union[List, Tuple], b: Union[List, Tuple], dim: int):
+    seq_length = len(a)
```

Comment:
This means that if b is longer than a, we truncate b. Is this what we want?

Reply:
```concat_tensor_sequence``` is being used in the ```BidirectionalDPLSTMLayer``` class and it takes the output sequences from the forward and reverse layer and so the lengths of two output sequences are the same. So, we are good for now. In case that we want to use this function for any other case, would you like me to throw some warning or exception in case ```a``` and ```b``` are of different lengths?

Reply:
Ah, that makes sense. How about renaming it to `_concat_sequence_directions`? We could also call `a` and `b` something like `forward` and `reverse` so that this is more explicit. At that point, the `if` check that they are of the same length would immediately make sense to the reader

Reply:
In addition to the above, I have changed the other two function names to have an underscore at the beginning in my recent commit.